### PR TITLE
=doc WIP #18471 examples for file uploads

### PR DESCRIPTION
**Pull request by johanandren**
Refs #18471

### CODE REVIEW

File: akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  //#simple-upload
+  val simpleRoute =
+    path("upload") {
+      entity(as[Multipart.FormData]) { formData =>
+        val completed: Source[File, _] = formData.parts.mapAsync(1) {
+          case BodyPart(name, entity, _, _) =>
+            val file = File.createTempFile("uploaded", "bin")
+            val sink = SynchronousFileSink(file)
+            entity.dataBytes.runWith(sink).map(_ => file)
+        }
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala]: Yes and no :smile: 

Yes that's the "simple to file" case, it should be documented that's true.
I also think this is too much ceremony and we should provide a directive for it, for the simple case `path()(simpleFileUpload(to:File)(onUploaded:Try[File] => Response))`

The simple one is important, but we also should document the awesome part – while the data is incoming do `dataBytes.via(Framing.delimiter("\n").map(readCsv).map(convert).to(Sink.actorRef(something))`

so it's *really streaming* and actually doing stuff to the data while its streaming :)

Comment (ktoso) [inline akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala]: Maybe `path()(simpleFileUpload(to:File)(onUploaded:Try[File] => Response))` as a separate ticket. 
But here we should do something with actual streaming, for sure.

File: akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+  //#simple-upload
+  val simpleRoute =
+    path("upload") {
+      entity(as[Multipart.FormData]) { formData =>
+        val completed: Source[File, _] = formData.parts.mapAsync(1) {
+          case BodyPart(name, entity, _, _) =>
+            val file = File.createTempFile("uploaded", "bin")
+            val sink = SynchronousFileSink(file)
+            entity.dataBytes.runWith(sink).map(_ => file)
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala]: also, `BodyBart("not-a-file", ) =>`
so we can for example get metadata about a movie...

Maybe that's good for the low-level case.

3 examples:
- low level, multiple bory parts - one metadata, one file
- simple file upload to file, as shown below
- streaming handling - split by line, handle some csv

File: akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    path("upload") {
+      entity(as[Multipart.FormData]) { formData =>
+        val partsSource: Source[Part, _] = formData.parts.mapAsync(1) { part =>
+          part.filename.fold[Future[Part]](
+            throw new RuntimeException("Expected all request parts to be file parts")) { fileName =>
+              if (fileName.endsWith(".bin")) saveFileToDisk(part)
+              else if (fileName.endsWith(".json")) processJsonPart(part)
+              else throw new RuntimeException(s"Unexpexted part suffix $fileName")
+            }
+        }
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala]: this example is a bit weird, let's not I think. Focus on the above one

File: akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala (hunk 2 above)
Comment (johanandren) [inline akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala]: But the metadata+body parts *is* what the below sample is (based on Johannes snippet from the mailing list).

The CSV stuff, is that related to file uploads or just a sample that should be in there, or you mean that the file will contain CSV that we will parse?

Comment (ktoso) [inline akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala]: I thought it simpler to show a csv here than to go all in with json parsing (which needs implicit context), for the sample we can just `via(framing).mapConcat(_.split(","))`

File: akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+          }
+
+          // when processing have finished create a response for the user
+          onSuccess(allPartsF) { allParts =>
+            complete {
+              "ok!"
+            }
+          }
+        }
+      }
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala]: I guess that's right, @jrudolph could you have a look here please?

File: akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    val splitLines = Framing.delimiter(ByteString("\n"), 256)
+
+    val csvUploads =
+      path("metadata" / LongNumber) { id =>
+        entity(as[Multipart.FormData]) { formData =>
+          val done = formData.parts.mapAsync(1) {
+            case b: BodyPart if b.filename.exists(_.endsWith(".csv")) =>
+              b.entity.dataBytes
+                .via(splitLines)
+                .map(_.utf8String.split(",").toVector)
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala]: let's add a `// naive implementation, could be a stream as well`

File: akka-docs-dev/rst/scala/http/routing-dsl/index.rst [hunk 6]
```diff
@@ -74,3 +74,25 @@ Failures and exceptions inside the Routing DSL
...
+
+File uploads
+^^^^^^^^^^^^
+
+Handling a simple file upload from for example a browser form with a `file` input can be done
+by accepting a `Multipart.FormData` entity, note that the body parts are `Source` rather than
+all available right away, and so is the individual body part payload so you will need to consume
+those streams both for the file and for the form fields.
+
+Here is a simple example which just dumps the uploaded file into a temporary file on disk, collects
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/http/routing-dsl/index.rst]: "which streams the uploaded bytes into a file"? Need to stress that it's still streaming (people were confused when I showed a file in an example)

**Comment by ktoso** [comment]
Looks good from my perspective if it works (curl it) :-)
We need a nice directive for streaming fileuploads, something that looks nice as a oneliner - would you like to pick that one up next?

**Comment by johanandren** [comment]
Yeah, that would be a good bite sized chunk to work on.

**Comment by ktoso** [comment]
:-)

File: akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+  "simple-upload" in {
+    val uploadVideo =
+      path("video") {
+        entity(as[Multipart.FormData]) { formData =>
+
+          // collect all parts of the multipart as it arrives into a map
+          val allPartsF: Future[Map[String, Any]] = formData.parts.mapAsync[(String, Any)](1) {
+
+            case b: BodyPart if b.name == "file" =>
```
Comment (drewhk) [inline akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala]: we should probably have a directive for storing things in a temp file.

**Comment by drewhk** [comment]
LGTM

File: akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala (hunk 7 above)
Comment (ktoso) [inline akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/FileUploadExamplesSpec.scala]: Yup, definitely needed - I'll see if we have a ticket and if not open one.

**Comment by ktoso** [comment]
Let's add the javadsl side of things as part of this PR, and the high level directive in another PR :)

**Comment by johanandren** [comment]
Java side of the DSL not done yet #15674

**Comment by ktoso** [comment]
Ouch right, sorry for leading you into another incomplete API!
Squash and let's merge here.

**Comment by ktoso** [comment]
I'm going to merge this and we can add the "simple" APIs here along in the PR about them.

**Comment by ktoso** [comment]
Nice job, thanks @johanandren !

**Pull request merged**